src/ConnectionAutomaton.h

```diff
@@ -50,7 +50,8 @@ class ConnectionAutomaton :
     /// Registered as an input in the DuplexConnection.
     public Subscriber<std::unique_ptr<folly::IOBuf>>,
     /// Receives signals about connection writability.
-    public Subscription {
+    public Subscription,
+    public std::enable_shared_from_this<ConnectionAutomaton> {
```

Comment:
Thanks for the comment here. I found it also confusing when I am using classes deriving from enable_shared_from_this and allocate them on the stack and then be surprised it is throwing execptions about bad weak_ptr. I was going to do what you are suggesting but I didn't like the duplicated code for the duplicated creator methods.
Having the API accepting only shared_ptrs of objects, the user is forced to allocate them all on heap so I never fell in the trap of bad weak_ptr in any of the unit tests.